lib/spack/spack/stage.py

```diff
@@ -725,6 +737,138 @@ def _add_to_root_stage(self):
+        sub_dir = self._compute_next_open_subdir()
+        full_subdir = self._remote_root / sub_dir
+        full_subdir.mkdir()
+        return full_subdir
+
+    def _compute_next_open_subdir(self, last=None):
+        def inc(c):
+            curr = ord(c[-1])
+            over = curr // 122
+            if over:
+                return (inc(c[:-1]) if len(c[:-1]) else "a") + "a"
+            return c[:-1] + chr(curr + 1)
+
+        sort_key = lambda x: (len(x.name), x.name)
+        current_ext_stages = list(self._remote_root.iterdir())
+        if not current_ext_stages:
+            # no currently extant external stages, start enumerating with 'a'
+            return "a"
+        last = sorted(current_ext_stages, key=sort_key)[-1]
+        return inc(last.name)
+
+    def _teardown_remote_build_dir(self):
+        """Destroy external build tree if not keep-stage
+        Otherwise this is kept as usual"""
+
+        def teardown(pth: Path):
+            for sub_item in pth.iterdir():
+                if sub_item.is_dir():
+                    teardown(sub_item)
+                else:
+                    sub_item.unlink()
+            pth.rmdir()
+
+        if self._remote_build_dir.exists():
+            teardown(self._remote_build_dir)
+        self._remote_build_dir = None
+
+    def _destroy_remote(self):
+        # copy back to stage may fail in event of error, make sure we clean up the
+        # associated external build dir in that event unless we're keeping the
+        # parent stage on cleanup
+        # If remote stage is not set, we never created one, package is already
+        # installed and we should do nothing here
+        if self._remote_build_dir:
+            try:
+                self._remove_context_link()
+            finally:
+                self._teardown_remote_build_dir()
+
+    def contains(self, file_path):
+        cont = super(CMakeBuildStage, self).contains(file_path)
+        if not cont:
+            return str(self._remote_build_dir) in file_path
+        return cont
+
+    def create(self):
+        if not self.created:
+            # Establish typical root stage
+            super(CMakeBuildStage, self).create()
+            # Now establish remote build stage
+            # or pick up extant stage from existing  symlink
+            if not self._root_stage_context.exists():
+                try:
+                    self._remote_build_dir = self._setup_remote_build_stage()
+                    self._establish_context_link()
+                except Exception:
+                    self._teardown_remote_build_dir()
+                    self.created = False
+                    raise
+            else:
+                self._remote_build_dir = self._root_stage_context.resolve()
+
+    def destroy(self):
+        self._destroy_remote()
+        super(CMakeBuildStage, self).destroy()
+
+    def path_rel_to_stage(self, glob_expr):
+        return os.path.relpath(glob_expr, self._remote_build_dir)
```

Comment:
We spoke about how this avoids strange issues given that `glob_expr` would potentially be an absolute path that resides inside the `build_directory` - in this case distinct from the stage path.

I think we also briefly discussed the possibility that a user could request archiving a file that is not inside the `build_directory` (I know of no such case, but we don't prevent it). Does this logic address that somehow? If it doesn't and you don't want to consider that case, I think it's acceptable to raise an exception if the glob_expr is not a subdir of `_remote_build_dir`.

Reply:
Oh and this is complex enough that IMO a docstring would be useful (i.e. this function can explain why it's calculating relative to this more-specific property)

Reply:
I've added functionality to fall back on the default (pre `CMakeBuildStage`) behavior when an archive file falls outside of the scope of the `build_directory`, so that behavior will not change, as well as a relevant docstring.